add(tutorial): Crossplane doc MTA-2954 (#1720)
scaleway/docs-content: tutorials/get-started-crossplane-kubernetes/index.mdx

---
meta:
 title: Getting started with Crossplane provider on Scaleway Kubernetes
 description: This page shows you how to get started with the Crossplane provider using Kubernetes clusters
content:
 h1: Getting started with Crossplane provider on Scaleway Kubernetes
 paragraph: This page shows you how to get started with the Crossplane provider using Kubernetes clusters
tags: crossplane provider kubernetes cluster
categories:
 - kubernetes
dates:
 validation: 2023-05-05
 posted: 2023-05-05
---

Created by Upbound, [Crossplane](crossplane.io) is a Kubernetes-based platform that allows you to manage resources through a unified control plane. `crossplane-provider-scaleway` is a Crossplane provider that helps you manage Scaleway cloud resources in Kubernetes. It is built using [Upjet](https://github.com/upbound/upjet) code generation tools and it exposes managed resources for Scaleway that adhere to Crossplane's eXtensible Resource Model (XRM). As a model, XRM ensures that users can easily and consistently manage cloud resources across different cloud providers.

> In this tutorial, you will learn how to install Upbound Universal Crossplane (UXP) into your Kubernetes cluster, configure the provider to authenticate with Scaleway, and create a Scaleway [Kubernetes](https://www.scaleway.com/en/docs/containers/kubernetes/concepts/) managed resource directly from your cluster.

Author: Bien que contre intuitif, c'est bien l'acronyme officiel utilisé! X = Cross

<Macro id="iam-requirements" />

<Message type="requirement">
 - You have an account and are logged into the [Scaleway console](https://console.scaleway.com)
 - You have created a [Kubernetes cluster](/en/docs/containers/kubernetes/how-to/create-cluster/) and downloaded its [Kubeconfig file](https://www.scaleway.com/en/docs/containers/kubernetes/concepts#kubeconfig)
 - You have [created an API key](/en/docs/console/my-project/how-to/generate-api-key)
 - You have [installed curl](https://curl.se/download.html)
 - You have [installed kubectl](https://github.com/kubernetes/kops/blob/master/docs/install.md) on your local computer
</Message>

<Message type="note">
   All commands use the current `kubeconfig` context and configuration.
</Message>

## Installing the Up command-line

> Run the following command to download and install the Upbound `up` command-line:

Reviewer: ```suggestion
Run the following command to download and install the Upbound `up` command-line interface:
```

> ````

Reviewer 2: You only need 3 ``` everywhere, I don't think it would render well in the preview if you add 4

curl -sL "https://cli.upbound.io" | sh
sudo mv up /usr/local/bin/

> ````

Reviewer 2: same comment as above and for the following steps too

<Message type="note">
   This tutorial uses `curl` but other command-line tools are available. You can refer to the [official dedicated documentation](https://docs.upbound.io/cli/#install-the-up-command-line) for more information.
</Message>

## Installing Upbound Universal Crossplane

Open a terminal and run the following `up uxp install` command to install [Upbound Universal Crossplane (UXP)](https://docs.upbound.io/uxp/):

> ````

Reviewer: All these must be indented otherwise it displays all steps as step 1

up uxp install
UXP 1.9.0-up.3 installed
````

## Installing the provider into your Kubernetes cluster

1. Run the following command to create a `.kube` directory in which you will store your `kubeconfig` file:

````
mkdir .kube
````

2. Run the following command to open the `config` file in your `.kube` directory. The file should be empty.

````
nano .kube/config
````

> 3. Paste the content of your Kubeconfig file and save the changes by following the information displayed on your terminal. Here is an example on a Mac Terminal screen.

Reviewer 2: ```suggestion
3. Paste the content of your Kubeconfig file and save the changes by following the information displayed on your terminal. Here is an example on a Mac terminal:
```

<Lightbox src="scaleway-kubeconfig-terminal-screen.webp" alt="" />

4. Run the following command to install the provider into the Kubernetes cluster using the previously set Kubernetes configuration file.

````
cat <<EOF | kubectl apply -f -
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
 name: provider-scaleway
spec:
 package: xpkg.upbound.io/scaleway/provider-scaleway:v0.1.0
EOF
````

> You should get an output similar to this:

Reviewer 2: ```suggestion
You should get an output similar to the following:
```

````
provider.pkg.crossplane.io/provider-scaleway Created
````

5. Run the following command to verify whether the provider has been properly installed.

````
kubectl get providers
````

> You should get an output similar to this, with the value for `installed` returning `True`:

Reviewer 2: ```suggestion
You should get an output similar to the following, with the value for `installed` returning `True`:
```

````
NAME                INSTALLED   HEALTHY   PACKAGE                                             AGE
provider-scaleway   True        True   xpkg.upbound.io/scaleway/provider-scaleway:v0.1.0        11s
````

<Message type="note">
   The procedure may take up to 5 minutes for `HEALTHY` to report `True`.
</Message>

> 6. If the `INSTALLED` field is empty, it means the process of downloading and installing the provider has failed. Run `kubectl describe providers` to get more information.

Reviewer 2: The first part of this step is not a step, maybe you could convert the step into into a "tip" or a "note" message box

````
nano secret.yaml
````

2. Edit the request payload you will use to create your Kubernetes secret configuration file. Replace the parameters in the following example using the information in the configuration reference table:

````
apiVersion: v1
kind: Secret
metadata:
 name: name-of-secret
 namespace: crossplane-system
type: Opaque
stringData:
 credentials: |
   {
     "access_key": "SCWXXXXXXXXXXXXXXXXX",
     "secret_key": "11111111-1111-1111-1111-111111111111",
     "project_id": "11111111-1111-1111-1111-111111111111",
     "region": "fr-par",
     "zone": "fr-par-1"
   }
````

<Message type="important">
   Secret names must be a valid DNS subdomain, meaning they must consist of lowercase alphanumeric characters, hyphens ("-") or periods and must start and end with an alphanumeric character. Underscores ("_") are not permitted.
</Message>

3. Paste the content of the request payload and save the changes by following the information displayed on your terminal.

4. Run the following command to create the secret previously defined in the secret.yaml file.

````
kubectl apply -f secret.yaml
````

> You should get an output similar to this:

Reviewer 2: ```suggestion
You should get an output similar to the following:
```

````
secret/name-of-secret created
````

5. Run the following command to create a `crossplane-system` [namespace](/en/docs/containers/container-registry/concepts/#namespace):

````
kubectl create namespace crossplane-system --dry-run=client -o yaml | kubectl apply -f -
````

> You should get an output similar to this:

Reviewer 2: ```suggestion
You should get an output similar to the following:
```

````
namespace/crossplane-system created
````

> 6. You can run the following command to retrieve details about your secret:

Reviewer 2: ```suggestion
6. Run the following command to retrieve details about your secret:
```

````
kubectl describe secret -n crossplane-system
````

<Message type="note">

>    Note that because the information is considered as sensitive, the actual secret data will not be returned.

Reviewer 2: ```suggestion
   As the information is considered sensitive, the actual secret data will not be returned.
```

</Message>

> ### Configuration reference

Reviewer 3: ici c'est un H3 orphan, pas moyen de mettre un h2 ?

| Parameter    | Description                                                                                                                                                                                           |
| :----------- | :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `name`       | The name of the Kubernetes secret object.                                                                                                                                                             |
| `namespace`  | The Kubernetes namespace the secret is in.                                                                                                                                                            |
| `access_key` | The first part of your [API key](https://console.scaleway.com/iam/api-keys). It is a like a unique ID or username and not a sensitive piece of information.                                           |

> | `secret_key` | The second part of your [API key](https://console.scaleway.com/iam/api-keys). It is a unique ID or password and thus a sensitive piece of information to authenticate your Access Key.                |

Reviewer 2: ```suggestion
| `secret_key` | The second part of your [API key](https://console.scaleway.com/iam/api-keys). It is a unique ID or password and thus a sensitive piece of information to authenticate your access key.                |
```

## Creating a providerconfig file

Create a `.yaml` configuration file to attach your Scaleway credentials to the previously installed provider.

1. Run the following command to create a `providerconfig` file:

````
nano providerconfig
````

2. Edit the request payload you will use to create your `providerconfig` file. Replace the parameters in the following example using the information in the configuration reference table:

````
apiVersion: scaleway.upbound.io/v1beta1
kind: ProviderConfig
metadata:
 name: default
spec:
 credentials:
   source: Secret
   secretRef:
     name: name-of-secret
     namespace: crossplane-system
     key: credentials
````

3. Paste the content of the request payload and save the changes by following the information displayed on your terminal.

4. Run the following command to apply the previously set configuration to the secret. Make sure to replace the example file path with your own.

````
kubectl apply -f /users/FileName/providerconfig
````

> You should get an output similar to this:

Reviewer 2: ```suggestion
You should get an output similar to the following:
```

````
providerconfig.scaleway.upbound.io/default created
````

> 5. Run the following command to verify whether contents of the file have been successfully applied to the Kubernetes cluster.

Reviewer 2: ```suggestion
5. Run the following command to verify whether contents of the file have been applied to the Kubernetes cluster.
```

````
kubectl describe providerconfigs
````

> You should get an output similar to this:

Reviewer 2: ```suggestion
You should get an output similar to the following:
```

 Resource Version:  5471213446
 UID:               46460c35-b937-4459-8972-9fba7e211348
Spec:
 Credentials:
   Secret Ref:
     Key:        credentials
     Name:       name-of-secret
     Namespace:  crossplane-system
   Source:       Secret
Status:
Events:  <none>
````

### Configuration reference

| Parameter   | Description                                |
| :---------- | :----------------------------------------- |
| `namespace` | The Kubernetes namespace the secret is in. |
| `name`      | The name of the Kubernetes secret object.  |

### Creating a managed resource

1. Run the following command to create a Scaleway Object Storage bucket:

````
nano bucket.yaml
````
2. Paste the following example and save the changes by following the information displayed on your terminal:

````
apiVersion: object.scaleway.upbound.io/v1alpha1
kind: Bucket
metadata:
 name: object-bucket
spec:
 forProvider:
   name: crossplane-object-bucket
 providerConfigRef:
   name: default
````

3. Run the following command to apply the previously set configuration to the bucket. Make sure to replace the example file path with your own.

````
kubectl apply -f bucket.yaml
````

> You should get an output similar to this:

Reviewer 2: ```suggestion
You should get an output similar to the following:
```

````
bucket.object.scaleway.upbound.io/object-bucket created
````

4. Run the following command to retrieve details on your newly created bucket:

````
kubectl get buckets
````

> You should get an output similar to this, with the values for `READY` and `SYNCED` returning `True`:

Reviewer 2: ```suggestion
An output similar to the following should display, with the values for `READY` and `SYNCED` returning `True`:
```

````
NAME                           READY   SYNCED   EXTERNAL-NAME                     AGE
object-bucket                  True    True     fr-par/crossplane-object-bucket   9s
````

### Deleting a managed resource

1. Run the following command to delete the bucket you just created:

````
kubectl delete -f bucket.yaml
````

> You should get an output similar to this, confirming the bucket deletion:

Reviewer 2: ```suggestion
You should get an output similar to the following, confirming the bucket deletion:
```

````
bucket.object.scaleway.upbound.io "object-bucket" deleted
````

2. Run the following command to verify whether the bucket was properly deleted:

````
kubectl get buckets
````

> You should get an output similar to this:

Reviewer 2: ```suggestion
You should get an output similar to the following:
```
